### PR TITLE
Slack interceptor 1542

### PR DESCRIPTION
**Pull request by ilan-pinto**
# Changes

Added a Slack `Interceptor` that allows you to extract fields from a slack slash command [payload](https://api.slack.com/interactivity/slash-commands#app_command_handling) which are sent in the  http form-data section.

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [V] Includes [tests](https://github.com/tektoncd/community/blob/master/standards.md#tests) (if functionality changed/added)
- [V] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#docs) (if user facing)
- [V] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commits)
- [V] Release notes block has been filled in or deleted (only if no user facing changes)

_See [the contribution guide](https://github.com/tektoncd/triggers/blob/master/CONTRIBUTING.md) for more details._

# Release Notes

```release-note

Added a Slack `Interceptor` that allows you to extract fields from a slack slash command [payload](https://api.slack.com/interactivity/slash-commands#app_command_handling) which are sent in the  http form-data section.

```

### CODE REVIEW

**Review by khrm — COMMENTED** [review]
/ok-to-test

**Comment by ilan-pinto** [comment]
@khrm how can i fix the failing test ?

It seems to be coming from the framework

**Review by khrm — COMMENTED** [review]
I think you are using go1.18 or go1.20. While using scripts in hack folder, you need to use go1.19 version because test scripts uses go1.19 version.

**Comment by ilan-pinto** [comment]
@khrm fixed conflicts

File: pkg/sink/sink.go [hunk 1]
```diff
@@ -554,7 +558,28 @@ func (r Sink) ExecuteInterceptors(trInt []*triggersv1.TriggerInterceptor, in *ht
 		}
 		// Clear interceptorParams for the next interceptor in chain
 		request.InterceptorParams = map[string]interface{}{}
+
+		// parse form-data payload for slack only
+		if v := in.Header.Get("X-Slack-Signature"); v != "" {
```
Comment (dibyom) [inline pkg/sink/sink.go:563]: can we do this generally for all urlencoded form bodies? i.e. if the content type is urlencoded, we can parseQuery and use the jsonified string as the body?

Comment (ilan-pinto) [reply pkg/sink/sink.go:563]: yes

**Review by khrm — COMMENTED** [review]
/approve

We need to mention that payload verification needs to be disabled for slack interceptor.

File: pkg/sink/sink.go [hunk 2]
```diff
@@ -554,7 +558,24 @@ func (r Sink) ExecuteInterceptors(trInt []*triggersv1.TriggerInterceptor, in *ht
 		}
 		// Clear interceptorParams for the next interceptor in chain
 		request.InterceptorParams = map[string]interface{}{}
+
+		// check if string is urlencoded
```
Comment (dibyom) [inline pkg/sink/sink.go:562]: The logic here looks fine but shouldn't we call it when we first create the request object around line 467?

File: docs/interceptors.md [hunk 3]
```diff
@@ -474,6 +475,38 @@ spec:
...
+          params:
+            - name: requestedFields
+              value: 
+                - text   
+```
+Note: payload-validation must be disabled by add adding the following annotation 
+
+```yaml
+  annotations:
+    tekton.dev/payload-validation: "false"
```
[the remark this answers is not in the document]
Comment (dibyom) [reply docs/interceptors.md:506]: cc @khrm

File: test/e2e-tests-examples.sh [hunk 4]
```diff
@@ -76,19 +76,20 @@ apply_files() {
...
   }
+  
   kubectl get eventlisteners
 }
 
 ignoreExamples=( cron trigger-ref triggergroups )
 
 port_forward_and_curl() {
-
+sleep 10
```
[the remark this answers is not in the document]
Comment (ilan-pinto) [reply test/e2e-tests-examples.sh:92]: I faced issues on my local when using kind. it took more than 3 sec for the event listener to be available

File: pkg/apis/triggers/v1beta1/trigger_types.go [hunk 5]
```diff
@@ -155,6 +155,13 @@ type WebhookInterceptor struct {
 	Header []v1beta1.Param `json:"header,omitempty"`
 }
 
+type SlackInterceptor struct {
+	// the Requested fields to be extracted from data form
+
+	// +listType=atomic
+	RequestedFields []string `json:"requestedFields,omitempty"`
```
[the remark this answers is not in the document]
Comment (ilan-pinto) [reply pkg/apis/triggers/v1beta1/trigger_types.go:162]: We extract only the requested fields. 
if wont do any work if nothing is specified

File: pkg/sink/sink.go (hunk 2 above)
Comment (ilan-pinto) [reply pkg/sink/sink.go]: @dibyom, good call. It simplified the Slack interceptor. 
All tests are ok on my local but for some reason [this](https://github.com/ilan-pinto/triggers/blob/79a1e8fd69901eb2baaf37b3f00f5b828e979c54/test/eventlistener_test.go#L127) test fails in GitHub action

File: pkg/interceptors/slack/slack.go [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+
+var _ triggersv1.InterceptorInterface = (*Interceptor)(nil)
+
+type Interceptor struct {
+	SecretGetter interceptors.SecretGetter
+}
+
+// Interceptor parses all the requests fields from the slack form-data request
+// and adds them to the extension
+func (*Interceptor) Process(ctx context.Context, r *triggersv1.InterceptorRequest) *triggersv1.InterceptorResponse {
```
Comment (khrm) [inline pkg/interceptors/slack/slack.go:36]: Can you put  comments to disable parameter checking with revive linter.

```
// revive:disable:unused-parameter

func (*Interceptor) Process(ctx context.Context, r *v1beta1.InterceptorRequest) *v1beta1.InterceptorResponse {
....
}

// revive:enable:unused-parameter
```

**Review by khrm — COMMENTED** [review]
/lgtm

**Review by khrm — COMMENTED** [review]
/kind feature

**Pull request merged**